### PR TITLE
[MLIR][OpenMP] Support lowering of host_eval to LLVM IR

### PR DESCRIPTION
**Pull request by skatrak**
This patch updates the MLIR to LLVM IR lowering of `omp.target` to support passing `num_teams`, `num_threads`, `thread_limit` and SPMD loop bounds through the `host_eval` argument of `omp.target`.

This replaces the previous implementation where this information was directly attached to the `omp.target` operation rather than captured to be used by the corresponding nested operation.

The implementation of `TargetOp::getInnermostCapturedOmpOp` is also improved to address bugs in the detection of target SPMD in MLIR, uncovered by changes to the translation to LLVM IR.

### CODE REVIEW

**Comment by skatrak** [comment]
PR stack:
- #178
- #179
- #180

File: mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp [hunk 1]
```diff
@@ -3635,6 +3636,63 @@ static uint64_t getTeamsReductionDataSize(mlir::omp::TeamsOp &teamsOp) {
   return getReductionDataSize<mlir::omp::TeamsOp>(teamsOp);
 }
 
+static void
```
Comment (bhandarkar-pranav) [inline mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:3639]: Could you please add a comment here? The name is self-explanatory so a smallish comment should be good enough

File: mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp [hunk 2]
```diff
@@ -3635,6 +3636,63 @@ static uint64_t getTeamsReductionDataSize(mlir::omp::TeamsOp &teamsOp) {
...
+extractHostEvalClauses(OperandRange ops, llvm::ArrayRef<BlockArgument> args,
+                       Value &numThreads, Value &numTeamsLower,
+                       Value &numTeamsUpper, Value &threadLimit,
+                       llvm::SmallVectorImpl<Value> *lowerBounds = nullptr,
+                       llvm::SmallVectorImpl<Value> *upperBounds = nullptr,
+                       llvm::SmallVectorImpl<Value> *steps = nullptr) {
+  for (auto item : llvm::zip_equal(ops, args)) {
+    Value op = std::get<0>(item), arg = std::get<1>(item);
```
Comment (bhandarkar-pranav) [inline mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:3647]: Nit: I generally refrain from making comments about variable names but in this case could you consider renaming `op` and `arg` to `hostEvalVar` and `blockArg` or something on these lines?

File: mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp [hunk 3]
```diff
@@ -3754,26 +3819,124 @@ static void initTargetDefaultBounds(
...
+  if (numThreads)
     bounds.MaxThreads = moduleTranslation.lookupValue(numThreads);
 
-  if (Value tripCount = targetOp.getTripCount())
-    bounds.LoopTripCount = moduleTranslation.lookupValue(tripCount);
+  // To calculate the trip count, we first create a placeholder set of canonical
+  // loops which we then collapse. Then, we skip execution of the collapsed loop
+  // and remove the basic blocks that originally defined it. The trip count
+  // remains available in the entry block.
+  // TODO: Improve implementation by extracting the logic to calculate the trip
```
Comment (bhandarkar-pranav) [inline mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:3863]: Pardon my ignorance, but is there anything preventing us from doing it this way right now?

**Review by bhandarkar-pranav — COMMENTED** [review]
Minor comments. I went over the entire PR of course, but I'd prefer if someone more familiar with the SPMD stuff went over it so we are sure it is good.

File: mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp (hunk 3 above)
Comment (skatrak) [reply mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp]: I wanted to avoid touching the implementation of canonical loops creation, since I wasn't too familiar with it, but this approach was pretty terrible. I just made the change, which doesn't change much of the canonical loop creation, so hopefully should be fine.

**Review by skatrak — COMMENTED** [review]
Thank you @bhandarkar-pranav for taking a look. I should have addressed all your concerns with the last update. Regarding your comment about finding someone else familiar with the SPMD stuff, I happen to also be the guy who made the previous implementation 😅, and unsurprisingly changes look good to myself! 😄

**Review by bhandarkar-pranav — APPROVED** [review]
@skatrak  - Thank you for the changes. LGTM.

File: llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp [hunk 4]
```diff
@@ -3968,26 +3968,21 @@ OpenMPIRBuilder::createCanonicalLoop(const LocationDescription &Loc,
...
-
+Value *OpenMPIRBuilder::calculateCanonicalLoopTripCount(
+    const LocationDescription &Loc, Value *Start, Value *Stop, Value *Step,
+    bool IsSigned, bool InclusiveStop, const Twine &Name) {
   // Consider the following difficulties (assuming 8-bit signed integers):
   //  * Adding \p Step to the loop counter which passes \p Stop may overflow:
   //      DO I = 1, 100, 50
   ///  * A \p Step of INT_MIN cannot not be normalized to a positive direction:
   //      DO I = 100, 0, -128
+  updateToLocation(Loc);
```
Comment (Meinersbur) [inline llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:3979]: The code below was meant to check basic errors before modifying anything.

It doesn't matter that much where it is, but it should not give the impression that the comment above would apply to `updateToLocation`.

**Review by Meinersbur — APPROVED** [review]
OpenMPIRBuilder changes LGTM

No test added?

File: llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp (hunk 4 above)
Comment (skatrak) [reply llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp]: Thanks for pointing this out. I moved the call to `updateToLocation` below the asserts, as it was before.

**Comment by skatrak** [comment]
> OpenMPIRBuilder changes LGTM
> 
> No test added?

Thank you @Meinersbur for taking a look. I added a test that is basically copied from an existing one, since that one used to check all corner cases related to the trip count calculation. I don't know if it makes sense to keep both as they are, but let me know what you think.

**Comment by skatrak** [comment]
> What I had in mind was a test where `getHostEvalVars()` is non-empty and thus the code-generation changes, or is first possible without crashing.
> 

Ah, that makes sense. Sorry for the misunderstanding! I just added an MLIR to LLVM IR translation test for these host evaluated clauses. I thought you were referring to the OMPIRBuilder refactoring because that's what you said you had reviewed 😅.

> `calculateCanonicalLoopTripCount` was already tested indirectly through `createCanonicalLoop`. These are unit-tests, and since `calculateCanonicalLoopTripCount` is not available as its own "unit", I think you can remove the equivalent ones for `createCanonicalLoop`.

What I just did was to remove the `CanonicalLoopBounds` test and replace it with the `CanonicalLoopTripCount` test that does the same checks of the trip count, but without creating and looking at the canonical loop structure (which is already done by `CanonicalLoopSimple` above). Let me know if that's not what you meant, so that I can update it again.

File: llvm/unittests/Frontend/OpenMPIRBuilderTest.cpp [hunk 5]
```diff
@@ -1444,17 +1443,8 @@ TEST_F(OpenMPIRBuilderTest, CanonicalLoopBounds) {
...
-    Expected<CanonicalLoopInfo *> LoopResult =
-        OMPBuilder.createCanonicalLoop(Loc, LoopBodyGenCB, StartVal, StopVal,
-                                       StepVal, IsSigned, InclusiveStop);
-    assert(LoopResult && "unexpected error");
-    CanonicalLoopInfo *Loop = *LoopResult;
-    Loop->assertOK();
-    Builder.restoreIP(Loop->getAfterIP());
-    Value *TripCount = Loop->getTripCount();
+    Value *TripCount = OMPBuilder.calculateCanonicalLoopTripCount(
+        Loc, StartVal, StopVal, StepVal, IsSigned, InclusiveStop);
```
Comment (Meinersbur) [inline llvm/unittests/Frontend/OpenMPIRBuilderTest.cpp:1447]: That's actually pretty nice

**Review by Meinersbur — APPROVED** [review]
LGTM, thanks
